conflicts: field accepts only conjunctions?

I was trying to turn depopts version constraints into conflicts:
I thought the equivalent of`"fileutils" {(>= "2.0.0") & (< "3.0")}` in conflicts would be its negation:
`"fileutils" {(>= "3.0") | (< "2.0.0")}`, but I get an error (and no install/remove works in opam until I fix the constraint):
```
[ERROR] fileutils ((>= 3.0) | (< 2.0.0)) is not a valid conjunction
```

If I replace the `|` with `&` I get the behaviour I want (`"fileutils" {(>= "3.0") & (< "2.0.0")}`), however you can no longer evaluate the version constraint the "obvious" way (i.e. put your current version into the formula and evaluate it as a boolean formula): ((x >= "3.0") & (x < "2.0.0") is always false.

Do I understand it correctly the semantics of this constraint:
```
conflicts: [
 "package1" { (>= "3.0") & (< "1.0") } 
 "package2" { (>= "4.0") } 
] 
```

to mean that: there is a conflict when `version(package1) >= 3.0` and there is a conflict when `version(package1) < 1.0` and there is a conflict when `version(package2) >= 4.0`, however the `and` in this sentence is not a logical and, there is a conflict when any of the statements is true.
The formula to express when there is a conflict would actually be: `version(package1) >= 3.0 | version(package2) < 1.0 | version(package2) >= 4.0`.

--- a/admin-scripts/depopts_to_conflicts.ml
+++ b/admin-scripts/depopts_to_conflicts.ml
@@ -60,9 +60,9 @@ iter_packages ~opam:(fun _ opam ->
           depopts [] in
       let conflicts = OpamFile.OPAM.conflicts opam in
       let add_conflicts =
-        let c = OpamFormula.to_conjunction conflicts in
+        let c = OpamFormula.to_disjunction conflicts in
         List.filter (fun f -> not (List.mem f c)) neg_depopts in
-      OpamFormula.ands (conflicts :: [OpamFormula.of_conjunction add_conflicts])
+      OpamFormula.ands (conflicts :: [OpamFormula.of_disjunction add_conflicts])
     in
     let opam = OpamFile.OPAM.with_depopts opam depopts in
     let opam = OpamFile.OPAM.with_conflicts opam conflicts in

--- a/admin-scripts/to_1_1.ml
+++ b/admin-scripts/to_1_1.ml
@@ -115,7 +115,7 @@ let to_1_1 _ opam =
   let opam =
     OF.with_conflicts opam
       (OpamFormula.map (fun (n, cstr) ->
-           OpamFormula.Atom (n, rewrite_constraint ~conj:true cstr))
+           OpamFormula.Atom (n, rewrite_constraint ~conj:false cstr))
           (OF.conflicts opam))
   in
   let opam = OF.with_available opam (filter_vars (OF.available opam) +! FBool true) in

--- a/opam
+++ b/opam
@@ -8,7 +8,7 @@ authors: [
 ]
 homepage: "https://opam.ocaml.org"
 bug-reports: "https://github.com/ocaml/opam/issues"
-dev-repo: "https://github.com/ocaml/opam"
+dev-repo: "https://github.com/ocaml/opam.git"
 build: [
   ["./configure"]
   [make]

--- a/src/format/opamFile.ml
+++ b/src/format/opamFile.ml
@@ -1331,10 +1331,10 @@ module X = struct
           @ listm   t.build_test    s_build_test    OpamFormat.make_command
           @ listm   t.build_doc     s_build_doc     OpamFormat.make_command
           @ listm   t.remove        s_remove        OpamFormat.make_command
-          @ formula t.depends       s_depends       OpamFormat.make_ext_formula
-          @ formula t.depopts       s_depopts       OpamFormat.make_opt_formula
+          @ formula t.depends       s_depends       OpamFormat.(make_formula `Conj make_ext_constraints)
+          @ formula t.depopts       s_depopts       OpamFormat.(make_formula `Disj make_ext_constraints)
           @ option  t.depexts       s_depexts       OpamFormat.make_tags
-          @ formula t.conflicts     s_conflicts     OpamFormat.make_formula
+          @ formula t.conflicts     s_conflicts     OpamFormat.(make_formula `Disj make_constraints)
           @ list    t.features      s_features      OpamFormat.make_features
           @ list    t.libraries     s_libraries     OpamFormat.make_libraries
           @ list    t.syntax        s_syntax        OpamFormat.make_libraries
@@ -1424,7 +1424,7 @@ module X = struct
       let depends =
         assoc_default OpamFormula.Empty s s_depends
           (fun v ->
-             OpamFormat.parse_ext_formula v |>
+             OpamFormat.(parse_formula `Conj parse_ext_constraints v) |>
              check_depflags ~pos:(OpamFormat.value_pos v)) in
       let depopts =
         let rec cleanup ~pos acc disjunction =
@@ -1454,7 +1454,7 @@ module X = struct
         in
         assoc_default OpamFormula.Empty s s_depopts @@ fun value ->
         let f =
-          OpamFormat.parse_opt_formula value |>
+          OpamFormat.(parse_formula `Disj parse_ext_constraints) value |>
           check_depflags ~pos:(OpamFormat.value_pos value) in
         if not conservative &&
            not OpamFormatConfig.(!r.skip_version_checks) &&
@@ -1466,8 +1466,35 @@ module X = struct
             |> OpamFormula.ors
         else f
       in
-      let conflicts = assoc_default OpamFormula.Empty s s_conflicts
-          OpamFormat.parse_formula in
+      let conflicts =
+        let is_disjunction f =
+          List.for_all (function Atom _ -> true | _ -> false)
+            OpamFormula.(ors_to_list (to_atom_formula f))
+        in
+        let force_disjunction f =
+          OpamFormula.map_formula (function
+              | And (a, b) -> Or (a, b)
+              | f -> f)
+            f
+        in
+        let cleanup ~pos f =
+          if is_disjunction f then f else
+          (if OpamVersion.compare opam_version (OpamVersion.of_string "1.3") >= 0
+           then
+             OpamConsole.warning
+               "At %s:\n\
+                Conflicts must be a disjunction, '&' is not supported \
+                (treated as '|')."
+               (string_of_pos pos);
+           f
+           |> OpamFormula.map (fun (n,cs) -> Atom (n, force_disjunction cs))
+           |> force_disjunction)
+        in
+        assoc_default OpamFormula.Empty s s_conflicts @@ fun value ->
+        OpamFormat.(parse_formula `Disj parse_constraints value) |> fun f ->
+        if conservative then f
+        else cleanup ~pos:(OpamFormat.value_pos value) f
+      in
       let features = OpamFormat.assoc_default [] s s_features
           OpamFormat.parse_features in
       let libraries = assoc_list s s_libraries OpamFormat.parse_libraries in
@@ -1813,10 +1840,15 @@ module X = struct
            (not (OpamPackage.Name.Set.is_empty undep_pkgs)));
         cond 42 `Error
           "The 'dev-repo' field doesn't specify an explicit VCS. You may use \
-           URLs of the form \"git+https://\""
+           URLs of the form \"git+https://\" or a \".hg\" or \".git\" suffix"
           (match t.dev_repo with
            | None | Some (Git _ | Darcs _ | Hg _) -> false
-           | Some (Version _ | Local _ | Http _) -> true)
+           | Some (Version _ | Local _ | Http _) -> true);
+        cond 43 `Error
+          "Conjunction used in 'conflicts:' field. Only '|' is allowed"
+          (OpamVersion.compare t.opam_version (OpamVersion.of_string "1.3") >= 0 &&
+           List.exists (function Atom _ -> false | _ -> true) @@
+           OpamFormula.(ors_to_list (to_atom_formula t.conflicts)));
       ]
       in
       OpamStd.List.filter_map (fun x -> x) warnings
@@ -2356,9 +2388,9 @@ module X = struct
       let build = safe [] OpamFormat.assoc_list s s_build OpamFormat.parse_commands in
       let env = safe [] OpamFormat.assoc_list s s_env
           (OpamFormat.parse_list_list OpamFormat.parse_env_variable) in
-      let packages = safe OpamFormula.Empty 
+      let packages = safe OpamFormula.Empty
           (OpamFormat.assoc_default OpamFormula.Empty)
-          s s_packages OpamFormat.parse_formula in
+          s s_packages OpamFormat.(parse_formula `Conj parse_constraints) in
       let preinstalled = safe false
           (OpamFormat.assoc_default false) s s_preinstalled OpamFormat.parse_bool in
       let tags = assoc_string_list s s_tags in
@@ -2412,7 +2444,7 @@ module X = struct
             make_variable (s_configure, make_string_list s.configure);
             make_variable (s_make, make_string_list s.make);
             make_variable (s_build, make_commands s.build);
-            make_variable (s_packages, make_formula s.packages);
+            make_variable (s_packages, make_formula `Conj make_constraints s.packages);
             make_variable (s_env, make_list make_env_variable s.env);
             make_variable (s_tags, make_string_list s.tags);
           ] @ (

--- a/src/format/opamFormat.ml
+++ b/src/format/opamFormat.ml
@@ -478,25 +478,24 @@ let parse_package_version ?expected = function
   | x -> bad_format ~pos:(value_pos x) "Expected a package version"
 
 (* parse a list of formulas *)
-let rec parse_formulas opt ~constraints t =
+let rec parse_formulas join ~constraints t =
   let rec aux = function
     | String _ as t -> Atom (parse_package_name t, constraints [])
     | Option (_, t, g) -> Atom (parse_package_name t, constraints g)
-    | Group (_,g) -> Block (parse_formulas opt ~constraints (List (pos_null, g)))
+    | Group (_,g) -> Block (parse_formulas join ~constraints (List (pos_null, g)))
     | Logop (_, `Or, e1, e2) -> let left = aux e1 in Or (left, aux e2)
     | Logop (_, `And, e1, e2) -> let left = aux e1 in And (left, aux e2)
     | x ->
       bad_format ~pos:(value_pos x)
         "Expected a formula list of the form [ \"item\" {condition}... ]"
   in
-  OpamFormula.(if opt then ors else ands)
-    (List.map aux (lift_list t))
+  join (List.map aux (lift_list t))
 
-let rec make_formulas opt ~constraints t =
+let rec make_formulas split ~constraints t =
   let name = OpamPackage.Name.to_string in
   let rec aux = function
     | Empty             -> assert false
-    | Block f           -> Group (pos_null, lift_list (make_formulas opt ~constraints f))
+    | Block f           -> Group (pos_null, lift_list (make_formulas split ~constraints f))
     | And(e,f)          -> Logop (pos_null, `And, aux e, aux f)
     | Or(e,f)           -> Logop (pos_null, `Or, aux e, aux f)
     | Atom (n, cs)      ->
@@ -504,27 +503,22 @@ let rec make_formulas opt ~constraints t =
       | [] -> make_string (name n)
       | cs -> Option (pos_null, make_string (name n), cs)
   in
-  let to_list =
-    OpamFormula.(if opt then ors_to_list else ands_to_list) in
-  List (pos_null, List.map aux (to_list t))
+  List (pos_null, List.map aux (split t))
 
-let make_formula =
-  make_formulas false ~constraints:make_constraints
 
-let make_ext_formula =
-  make_formulas false ~constraints:make_ext_constraints
+let make_formula kind constraints t =
+  let split = match kind with
+    | `Conj -> OpamFormula.ands_to_list
+    | `Disj -> OpamFormula.ors_to_list
+  in
+  make_formulas split ~constraints t
 
-let parse_formula =
-  parse_formulas false ~constraints:parse_constraints
-
-let parse_ext_formula =
-  parse_formulas false ~constraints:parse_ext_constraints
-
-let parse_opt_formula =
-  parse_formulas true ~constraints:parse_ext_constraints
-
-let make_opt_formula =
-  make_formulas true ~constraints:make_ext_constraints
+let parse_formula kind constraints t =
+  let join = match kind with
+    | `Conj -> OpamFormula.ands
+    | `Disj -> OpamFormula.ors
+  in
+  parse_formulas join ~constraints t
 
 let parse_compiler_version = function
   | Ident (_,v)

--- a/src/format/opamFormat.mli
+++ b/src/format/opamFormat.mli
@@ -193,23 +193,29 @@ val parse_package_name : ?expected:name -> value -> name
 
 val parse_package_version : ?expected:version -> value -> version
 
-(** Parse package formula where AND are implicit: [x y -> x & y] *)
-val parse_formula : value -> formula
+(** Parser for version constraints in formulas *)
+val parse_constraints: value list -> OpamFormula.version_formula
 
-(** Build a formula where AND are implicit. *)
-val make_formula : formula -> value
+(** Parser for version constraints in formula with dependency flags *)
+val parse_ext_constraints:
+  value list -> package_dep_flag list * OpamFormula.version_formula
 
-(** Parse an AND formula where constraints are extended with leading keywords *)
-val parse_ext_formula : value -> ext_formula
+(** Builder for version constraints in formulas *)
+val make_constraints: OpamFormula.version_formula -> value list
 
-(** Make an AND formula where constraints are extended with leading keywords *)
-val make_ext_formula : ext_formula -> value
+(** Builder for version constraints in formula with dependency flags *)
+val make_ext_constraints:
+  package_dep_flag list * OpamFormula.version_formula -> value list
 
-(** Parse optional package formula where OR are implicit: [x y -> x | y] *)
-val parse_opt_formula : value -> ext_formula
+(** Parse package formula as a [`Conj]unction or [`Disj]unction, using the given
+    parser for constraints *)
+val parse_formula :
+  [`Conj | `Disj] -> (value list -> 'a) -> value -> (name * 'a) generic_formula
 
-(** Build a formula where OR are implicit. *)
-val make_opt_formula : ext_formula -> value
+(** Build a formula as a [`Conj]unction or [`Disj]unction, building constraints
+    with the given function *)
+val make_formula :
+  [`Conj | `Disj] -> ('a -> value list) -> (name * 'a) generic_formula -> value
 
 (** Parse compiler versions *)
 val parse_compiler_version: value -> compiler_version

--- a/src/format/opamFormula.ml
+++ b/src/format/opamFormula.ml
@@ -305,7 +305,8 @@ let to_conjunction t =
   match to_dnf t with
   | []  -> []
   | [x] -> x
-  | _   -> OpamConsole.error_and_exit "%s is not a valid conjunction" (to_string t)
+  | _   ->
+    failwith (Printf.sprintf "%s is not a valid conjunction" (to_string t))
 
 let ands l = List.fold_left make_and Empty l
 
@@ -321,7 +322,8 @@ let to_disjunction t =
   match to_cnf t with
   | []  -> []
   | [x] -> x
-  | _   -> OpamConsole.error_and_exit "%s is not a valid disjunction" (to_string t)
+  | _   ->
+    failwith (Printf.sprintf "%s is not a valid disjunction" (to_string t))
 
 let ors l = List.fold_left make_or Empty l
 

--- a/src/solver/opamSolver.ml
+++ b/src/solver/opamSolver.ml
@@ -161,8 +161,9 @@ let opam2cudf universe ?(depopts=false) ~build version_map package =
   let conflicts =
     try OpamPackage.Map.find package universe.u_conflicts
     with Not_found -> Empty in
-  let conflicts = (* prevents install of multiple versions of the same pkg *)
-    (name, None)::OpamFormula.to_conjunction conflicts in
+  let conflicts =
+    (name, None) :: (* prevents install of multiple versions of the same pkg *)
+    OpamFormula.to_disjunction conflicts in
   let installed = OpamPackage.Set.mem package universe.u_installed in
   let base = OpamPackage.Set.mem package universe.u_base in
   let reinstall = match universe.u_action with

